Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp

```diff
@@ -127,33 +130,35 @@ GStreamerVideoEncoder::~GStreamerVideoEncoder()
     close();
 }
 
-void GStreamerVideoEncoder::encode(RawFrame&& frame, bool shouldGenerateKeyFrame, EncodeCallback&& callback)
+Ref<VideoEncoder::EncodePromise> GStreamerVideoEncoder::encode(RawFrame&& frame, bool shouldGenerateKeyFrame)
 {
-    gstEncoderWorkQueue().dispatch([frame = WTFMove(frame), shouldGenerateKeyFrame, encoder = m_internalEncoder, callback = WTFMove(callback)]() mutable {
+    return invokeAsync(gstEncoderWorkQueue(), [frame = WTFMove(frame), shouldGenerateKeyFrame, encoder = m_internalEncoder]() mutable {
         auto result = encoder->encode(WTFMove(frame), shouldGenerateKeyFrame);
         if (encoder->isClosed())
-            return;
+            return EncodePromise::createAndReject("Encoder closed"_s);
 
-        String resultString;
-        if (result)
-            encoder->harness()->processOutputSamples();
-        else
-            resultString = "Encoding failed"_s;
+        if (!result)
+            return EncodePromise::createAndReject("Encoding failed"_s);
 
-        encoder->postTask([weakEncoder = ThreadSafeWeakPtr { encoder.get() }, result = WTFMove(resultString), callback = WTFMove(callback)]() mutable {
-            auto encoder = weakEncoder.get();
-            if (!encoder || encoder->isClosed())
-                return;
+        encoder->harness()->processOutputSamples();
+        return EncodePromise::createAndResolve();
+    });
+}
 
-            callback(WTFMove(result));
-        });
+Ref<GenericPromise> GStreamerVideoEncoder::flush()
+{
+    return invokeAsync(gstEncoderWorkQueue(), [encoder = m_internalEncoder] {
+        encoder->flush();
+        return GenericPromise::createAndResolve();
     });
 }
 
-void GStreamerVideoEncoder::flush(Function<void()>&& callback)
+Ref<GenericPromise> GStreamerVideoEncoder::setRates(uint64_t bitRate, double frameRate)
 {
-    gstEncoderWorkQueue().dispatch([encoder = m_internalEncoder, callback = WTFMove(callback)]() mutable {
-        encoder->flush(WTFMove(callback));
+    return invokeAsync(gstEncoderWorkQueue(), [encoder = m_internalEncoder, bitRate, frameRate] {
+        encoder->setRates(bitRate, frameRate);
+        encoder->applyRates();
```

Comment:
`applyRates()` is called by `setRates()` already, so maybe no need to call it a second time here?

Reply:
Done